### PR TITLE
Removed observer classes

### PR DESCRIPTION
**Pull request by AtlantiaKing**
Removed the Observer class and the templated TObserver class
Added AddListener and RemoveListener methods to the Subject and TSubject class with member functions as parameters

### CODE REVIEW

File: Utils/Subject.h [hunk 1]
```diff
@@ -1,104 +1,270 @@
...
+			virtual void Call() const = 0;
+			virtual unsigned int GetId() const = 0;
+			virtual bool IsConst() const = 0;
+		};
+
+		template<typename T>
+		struct ImplMemberFunctionWrapper final : public MemberFunctionWrapper
+		{
+			ImplMemberFunctionWrapper(T* _pThis, void(T::* _pFunction)())
+				: pThis{ _pThis }, pFunction{ _pFunction }, id{ ReflectionUtils::GenerateTypenameHash<T>() }
```
Comment (LeeVangraefschepe) [inline Utils/Subject.h]: Initialization order will not be executed like this (same for Const class type).
id,
pThis,
pFunctuin

File: Utils/Subject.h [hunk 2]
```diff
@@ -1,104 +1,270 @@
...
+			}
 		}
 		void RemoveListener(void(*pCallback)())
 		{
-			const auto& it{ std::find(begin(m_Functions), end(m_Functions), pCallback) };
-			if (it != end(m_Functions)) m_Functions.erase(it);
+			std::erase(m_Functions, pCallback);
 		}
 
 		void Notify()
```
Comment (LeeVangraefschepe) [inline Utils/Subject.h]: Notify call can be set const now

File: Utils/Subject.h [hunk 3]
```diff
@@ -1,104 +1,270 @@
...
+			ImplMemberFunctionWrapper(T* _pThis, void(T::* _pFunction)())
+				: pThis{ _pThis }, pFunction{ _pFunction }, id{ ReflectionUtils::GenerateTypenameHash<T>() }
 			{
-				observer->OnSubjectDestroy();
 			}
-		}
+			virtual ~ImplMemberFunctionWrapper() = default;
 
-		void AddListener(Observer* observer)
+			unsigned int id{};
```
Comment (LeeVangraefschepe) [inline Utils/Subject.h]: All member variables publicly exposed.
I know its internal but you also made a GetId() function for a publicly exposed id member var.

File: Utils/Subject.h [hunk 4]
```diff
@@ -1,104 +1,270 @@
...
 		void AddListener(void(*pCallback)(const T&))
 		{
 			m_Functions.push_back(pCallback);
 		}
-		void RemoveListener(TObserver<T>* observer)
+		template<typename TClass>
+		void RemoveListener(TClass* pThis, void(TClass::* pFunction)(const T&))
+		{
+			constexpr unsigned int classId{ ReflectionUtils::GenerateTypenameHash<TClass>() };
+			for (int i{ static_cast<int>(m_pMemberFunctions.size()) - 1 }; i >= 0; --i)
```
Comment (LeeVangraefschepe) [inline Utils/Subject.h:216]: Cache vector size. Every for tick the size will get recalculated.

File: Networking/Leap/LeapClient.cpp [hunk 5]
```diff
@@ -50,7 +50,7 @@ leap::networking::LeapClient::LeapClient(int port, const std::string& serverIp,
         std::stringstream ss{};
         ss << "Error connecting to server: " << WSAGetLastError() << std::endl;
         Debug::LogError(ss.str());
-
```
Comment (LeeVangraefschepe) [inline Networking/Leap/LeapClient.cpp]: Clear out empty change (merge conflict will happen otherwise)

**Review by LeeVangraefschepe — CHANGES_REQUESTED** [review]
Good cleanup you removed all #include "Observer.h" from every header file it was used in :)
Also another amazing usage for the new reflection utils.
Check comments for the review.

File: Utils/Subject.h (hunk 4 above)
Comment (AtlantiaKing) [reply Utils/Subject.h:216]: Vector size is only used for initializing the for index

**Review by LeeVangraefschepe — APPROVED** [review]
Code feedback is applied.

**Pull request merged**